pkg/querier/worker/scheduler_processor.go

```diff
@@ -136,20 +137,45 @@ func (sp *schedulerProcessor) processQueriesOnSingleStream(workerCtx context.Con
 }
 
 // process loops processing requests on an established stream.
-func (sp *schedulerProcessor) querierLoop(c schedulerpb.SchedulerForQuerier_QuerierLoopClient, address string, inflightQuery *atomic.Bool) (err error) {
+func (sp *schedulerProcessor) querierLoop(execCtx context.Context, c schedulerpb.SchedulerForQuerier_QuerierLoopClient, address string, inflightQuery *atomic.Bool) (err error) {
 	// Build a child context so we can cancel a query when the stream is closed.
-	ctx, cancel := context.WithCancelCause(c.Context())
+	// Note that we deliberately don't use c.Context() here, as that is cancelled as soon as the gRPC client observes an error,
+	// but we don't always want to cancel queries if the scheduler stream reports an error (eg. if the scheduler crashed).
+	ctx, cancel := context.WithCancelCause(execCtx)
 	defer func() {
 		cancel(util.NewCancellationErrorf("query-scheduler loop in querier for query-scheduler %v terminated with error: %w", address, err))
 	}()
 
+	var queryComplete chan struct{}
+
+	waitForQuery := func(err error) {
+		select {
+		case <-queryComplete:
+			// Query is already complete, nothing to do.
+			return
+		default:
+			// Query is not complete.
+			level.Info(sp.log).Log("msg", "query-scheduler loop received non-cancellation error, waiting for inflight query to complete...", "err", err, "address", address)
+			<-queryComplete
+			level.Info(sp.log).Log("msg", "query-scheduler loop received non-cancellation error and inflight query is complete, continuing", "err", err, "address", address)
+		}
+	}
+
 	for {
 		request, err := c.Recv()
 		if err != nil {
+			// If the query was cancelled, we don't want to wait for it to complete: we want to cancel it, which will be
+			// handled by the deferred context cancellation above.
+			// If we got another kind of error (eg. scheduler crashed), continue processing the query.
+			if !grpcutil.IsCanceled(err) && queryComplete != nil {
+				waitForQuery(err)
```

Comment:
Below we call `c.Send()`. After this change, that call will fail while we're in the `waitForQuery()` because the gRPC stream already failed. It's not a big deal, but we'll log the error, which may be deceiving. I'm wondering if we should keep track of "gRPC stream failed" and then log differently the `c.Send()` error (I think it's safer to keep calling `c.Send()` anyway even if `Recv()` has failed, cause the error could also be transient we can't know).

Skip this comment if turns out to make code more complicated.

Reply:
Good idea, changed in f96b38f